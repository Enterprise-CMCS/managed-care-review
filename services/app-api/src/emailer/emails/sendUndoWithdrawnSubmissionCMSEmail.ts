import type { ContractType } from '../../domain-models'
import type {
    EmailConfiguration,
    EmailData,
    StateAnalystsEmails,
} from '../emailer'
import { pruneDuplicateEmails } from '../formatters'
import {
    renderTemplate,
    stripHTMLFromTemplate,
    parseEmailDataWithdrawSubmission,
    type RateForDisplayType,
} from '../templateHelpers'

export const sendUndoWithdrawnSubmissionCMSEmail = async (
    contract: ContractType,
    ratesForDisplay: RateForDisplayType[],
    stateAnalystsEmails: StateAnalystsEmails,
    config: EmailConfiguration
): Promise<EmailData | Error> => {
    const toAddresses = pruneDuplicateEmails([
        ...stateAnalystsEmails,
        ...config.dmcpSubmissionEmails,
        ...config.oactEmails,
        ...config.devReviewTeamEmails,
    ])

    const etaData = parseEmailDataWithdrawSubmission(
        contract,
        ratesForDisplay,
        config,
        'UNDO_WITHDRAW'
    )
    if (etaData instanceof Error) {
        return etaData
    }

    const template = await renderTemplate(
        'sendUndoWithdrawnSubmissionCMSEmail',
        etaData
    )

    if (template instanceof Error) {
        return template
    } else {
        return {
            toAddresses,
            replyToAddresses: [],
            sourceEmail: config.emailSource,
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }${etaData.contractName} status update`,
            bodyText: stripHTMLFromTemplate(template),
            bodyHTML: template,
        }
    }
}
