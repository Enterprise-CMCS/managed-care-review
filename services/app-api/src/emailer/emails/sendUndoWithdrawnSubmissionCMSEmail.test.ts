import {
    mockContract,
    testEmailConfig,
    testStateAnalystsEmails,
} from '../../testHelpers/emailerHelpers'
import { sendUndoWithdrawnSubmissionCMSEmail } from './sendUndoWithdrawnSubmissionCMSEmail'

const testContract = mockContract({
    id: 'test-contract-123',
    consolidatedStatus: 'RESUBMITTED',
    reviewStatusActions: [
        {
            updatedAt: new Date('2025-05-10'),
            updatedBy: {
                role: 'CMS_USER',
                email: 'zuko@example.com',
                givenName: 'Zuko',
                familyName: 'Hotman',
            },
            updatedReason: 'Test Undo Withdraw',
            dateApprovalReleasedToState: undefined,
            actionType: 'UNDER_REVIEW',
            contractID: 'test-contract-123',
        },
        {
            updatedAt: new Date('2025-04-10'),
            updatedBy: {
                role: 'CMS_USER',
                email: 'zuko@example.com',
                givenName: 'Zuko',
                familyName: 'Hotman',
            },
            updatedReason: 'Test Withdraw',
            dateApprovalReleasedToState: undefined,
            actionType: 'WITHDRAW',
            contractID: 'test-contract-123',
        },
    ],
})

const testRatesForDisplay = [
    {
        id: 'rate-123',
        rateCertificationName: 'rate-name-123',
    },
    {
        id: 'rate-321',
        rateCertificationName: 'rate-name-321',
    },
]
const stateAnalystEmails = () => [...testStateAnalystsEmails]

describe('sendUndoWithdrawnSubmissionCMSEmail', () => {
    it('CMS email contains correct toAddresses', async () => {
        const template = await sendUndoWithdrawnSubmissionCMSEmail(
            testContract,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.toAddresses).toEqual(
            expect.arrayContaining([
                ...stateAnalystEmails(),
                ...testEmailConfig().dmcpSubmissionEmails,
                ...testEmailConfig().oactEmails,
            ])
        )
    })

    it('renders overall email for undo withdraw contract as expected', async () => {
        const template = await sendUndoWithdrawnSubmissionCMSEmail(
            testContract,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template.bodyHTML).toMatchSnapshot()
    })
})

describe('sendUndoWithdrawnSubmissionCMSEmail error handling', () => {
    it('returns an error if contract consolidated status is not RESUBMITTED', async () => {
        const contractWithWrongStatus = {
            ...testContract,
            consolidatedStatus: 'WITHDRAWN' as const,
        }

        const template = await sendUndoWithdrawnSubmissionCMSEmail(
            contractWithWrongStatus,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Contract consolidated status should be RESUBMITTED'
        )
    })

    it('returns an error if review status actions is empty', async () => {
        const contractWithMissingReviewActions = {
            ...testContract,
            reviewStatusActions: [],
        }

        const template = await sendUndoWithdrawnSubmissionCMSEmail(
            contractWithMissingReviewActions,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Latest contract review action is not UNDER_REVIEW'
        )
    })

    it('returns an error if the latest contract review action is not UNDER_REVIEW', async () => {
        const contractWithWrongActionType = {
            ...testContract,
            reviewStatusActions: [
                {
                    updatedAt: new Date('2025-04-10'),
                    updatedBy: {
                        role: 'CMS_USER' as const,
                        email: 'zuko@example.com',
                        givenName: 'Zuko',
                        familyName: 'Hotman',
                    },
                    updatedReason: 'Test Withdraw',
                    dateApprovalReleasedToState: undefined,
                    actionType: 'WITHDRAW' as const,
                    contractID: 'test-contract-123',
                },
            ],
        }

        const template = await sendUndoWithdrawnSubmissionCMSEmail(
            contractWithWrongActionType,
            testRatesForDisplay,
            stateAnalystEmails(),
            testEmailConfig()
        )

        expect(template).toBeInstanceOf(Error)
        expect((template as Error).message).toBe(
            'Latest contract review action is not UNDER_REVIEW'
        )
    })
})
