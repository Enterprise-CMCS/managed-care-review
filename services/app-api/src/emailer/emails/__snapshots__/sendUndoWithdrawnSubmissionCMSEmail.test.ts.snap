// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sendUndoWithdrawnSubmissionCMSEmail > renders overall email for undo withdraw contract as expected 1`] = `
"<h1 style="font-weight: normal;">Submission MCR-MN-0004-SNBC status was updated to 'Submitted' by CMS</h1>
<b>Updated by:</b> zuko@example.com<br />
<b>Updated on:</b> 05/09/2025<br />
<b>Reason:</b> Test Undo Withdraw<br />
<b>Related rate certification(s):</b>
<ul>
  <li><a href="http://localhost/rates/rate-123">rate-name-123</a></li>
  <li><a href="http://localhost/rates/rate-321">rate-name-321</a></li>
</ul>
<a href="http://localhost/submissions/test-contract-123">View submission in MC-Review</a>
"
`;
